### ISSUE TITLE
Make Lwt.pick [] raise Invalid_argument

### ISSUE THREAD

**Issue by aantron** [ocsigen/lwt#557]
This issue is about adding an `if` to [`Lwt.pick`][pick], [`Lwt.choose`][choose], [`Lwt.npick`][npick], [`Lwt.nchoose`][nchoose], and [`Lwt.nchoose_split`][nchoose_split], so that they raise `Invalid_argument` when given the empty list.

[pick]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/src/core/lwt.ml#L2678-L2700
[choose]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/src/core/lwt.ml#L2656-L2676
[npick]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/src/core/lwt.ml#L2791-L2845
[nchoose]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/src/core/lwt.ml#L2732-L2787
[nchoose_split]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/src/core/lwt.ml#L2850-L2925

Right now, each one of them looks like

```ocaml
let pick ps =
  the_code
```

and we need to change that to

```ocaml
let pick ps =
  if ps = [] then
    raise (Pervasives.Invalid_argument
      "Lwt.pick [] would return a promise that is pending forever")
  else
    the_code
```

<br/>

### Background

This was suggested by @hcarty in #551 and #552, and we had a discussion there. You're welcome to add more points to the discussion :) Perhaps we have missed something.

<br/>

### Testing

After making the change, run `make test`. Some of the tests will, naturally, fail ([[1]][pick-test] [[2]][choose-test] [[3]][npick-test] [[4]][nchoose-test] [[5]][nchoose_split-test]). To fix them, they just need to be changed to expect the exception. For example:

```ocaml
test "empty" begin fun () ->
  try
    ignore (Lwt.pick []);
    Lwt.return false
  with Pervasives.Invalid_argument "your chosen text" ->
    Lwt.return true
end;
```

[pick-test]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/test/core/test_lwt.ml#L2516-L2519
[choose-test]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/test/core/test_lwt.ml#L1913-L1916
[npick-test]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/test/core/test_lwt.ml#L2622-L2625
[nchoose-test]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/test/core/test_lwt.ml#L1980-L1983
[nchoose_split-test]: https://github.com/ocsigen/lwt/blob/020600c2304c3830a6cf71e143c2c3eaeb77155b/test/core/test_lwt.ml#L2038-L2041

For more build instructions, see [`doc/CONTRIBUTING.md`](https://github.com/ocsigen/lwt/blob/master/doc/CONTRIBUTING.md). Otherwise, please feel free to ask!

<br/>

### The “breaking” label

This just means that we will leave the PR unmerged for a bit, while we warn users and give them a chance to stop us (#293, #453). Someone may have a good idea of why this shouldn't be done. We don't think this will actually break any code that isn't already broken, but we have to be careful and thorough.

You don't have to worry about warning users, waiting to merge the PR yourself, etc. :) This a job for maintainers, and part of the Lwt release process. So open a PR, we will have a quick and normal review and discussion, and let you know when all is good :) We will take the PR from there, so you can forget all about it and focus on the rest of life :)

**Comment by aantron** [comment]
I'd like to also suggest that if the PR indents all the code currently in `Lwt.pick` and the other functions, the indentation should go in a separate commit, whose message says that it is only a change to whitespace or indentation. That way, if someone is later looking through the history for when and how these functions were changed, they won't be stuck comparing all the lines only to find out that they were just indented. The commit message will tell them that :)

**Comment by timreinke** [comment]
Thanks for the excellent writeup, Anton.  I'd like to pick this up.  A quick question of style: do you consider a helper function like `let fail_on_empty ps msg = if ps = [] then Pervasives.invalid_arg msg` bad form? I can appreciate that there is little logic here and moving the check into a helper function can actually obfuscate the flow and lead to dead code if behavior changes in the future (the helper function may be left behind). I'm interested in your opinion here.

**Comment by aantron** [comment]
Thanks :)

In this case, I personally would indeed consider the helper bad form, because, as you suggest, there is very little gain, and it obfuscates what is going on.

In the meantime, the function you suggested made me realize that you don't need `else` and so don't need to indent any of the existing code, which is good. Thanks for that :)

Regarding dead code, we shouldn't have a problem with dead code in Lwt (except in elaborate situations), because we have the unused variable warning enabled during development builds, and warnings as errors. So if this function became dead due to simply not being mentioned, we would have to address that right away.

**Comment by aantron** [comment]
Addressed by @timreinke in #562 (thanks!).
